File: 0009-palindrome-number/0009-palindrome-number.py
class Solution:
    def isPalindrome(self, x: int) -> bool:
        # negative nums or nums ending with 0 (but not 0 itself) cannot be palindromes
        if x < 0 or (x % 10 == 0 and x != 0):
            return False
        
        # build reverted half with modulo and division
        reverted_half = 0
        while x > reverted_half:
            reverted_half = reverted_half * 10 + x % 10
            x //= 10
        
        # for even length: x == reverted_half
        # for odd length: x == reverted_half // 10 (skip middle digit)
        return x == reverted_half or x == reverted_half // 10
        

File: 0009-palindrome-number/0009-palindrome-number.java
class Solution {
    public boolean isPalindrome(int x) {
        if (x < 0 || (x % 10 == 0 && x != 0)){          // negative ints or ints ending with 0 (but not 0 itself) are not palindromes
            return false;
        }

        int reversed = 0;                               // store reversed digits of the the second half of x
        while (x > reversed) {                          // build reversed half until it matches or exceeds (for odd digits) remaining x
            reversed = (reversed * 10) + (x % 10);      // append last digit of x to reversed
            x /= 10;                                    // remove last digit from x
        }

        return x == reversed || x == reversed / 10;     // for even length: x == reversed; for odd length: x == reversed/10 (skip middle digit)
    }
}